### PR TITLE
Add e2e review

### PR DESCRIPTION
**Pull request by zhenik**

### CODE REVIEW

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.testcontainers.containers.wait.strategy.Wait;
+
+import static org.awaitility.Awaitility.await;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class DockerComposeIT {
+
+    public static final String JSON_SERVER_EXPOSED = "http://localhost:3000";
+    public static final String HTTP_PRODUCER_EXPOSED = "http://localhost:8080";
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: nit: I'd rephrase these 2 to something like: `HTTP_PRODUCER_BASE_URL`

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    @ClassRule
+    public static DockerComposeContainer environment =
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: nit: add comments to explain what will happen under-the-hood.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        RestAssured.given()
+                .contentType(ContentType.JSON)
+                .body(messageJsonRepresentation)
+                .post(HTTP_PRODUCER_EXPOSED + "/messages")
+                .then()
+                .statusCode(202);
+
+        await()
+                .atMost(70, TimeUnit.SECONDS)
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: 70 seconds seems like too much time to expect a response. between 5 and 10 seconds could be a better boundary.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                .untilAsserted(
+                        () -> {
+                            MessageJsonRepresentation jsonPresentation =
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: typo: `jsonPresentation` to `jsonRepresentation`.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private static void setZookeeperAndKafka() {
+        kafka = new KafkaContainer(CONFLUENT_PLATFORM_VERSION);
+        kafka.start();
+    }
+
+    private static void setSchemaRegistry() {
+        // get network alias to be able connect other container to env
+        KAFKA_BROKER_INSIDE_DOCKER_ENV = "PLAINTEXT://" + kafka.getNetworkAliases().get(0) + ":9092";
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: I think you can use the `kafka.getBootstrapServers()` method to get this.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    public static String JSON_SERVER_INSIDE_DOCKER_ENV;
+    public static String JSON_SERVER_EXPOSED;
+
+    public static GenericContainer httpProducer;
+    public static String HTTP_PRODUCER_EXPOSED;
+
+    public static GenericContainer httpMaterializer;
+
+    @BeforeClass
+    public static void start() {
```
Comment (jeqo) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: nit: I'd recommend to add all the docker container config here, instead of partitioned methods. Then you expose the configs on the same place and avoid using static variables.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java (hunk 5 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: KafkaContainer use under the hood `SocatContainer proxy`, which makes `kafka.getBootstrapServers()` exposed to localhost, but inside docker network it uses `PLAINTEXT://<network-alias>:9092`. I have `KAFKA_BROKER_INSIDE_DOCKER_ENV`, which is bootstrap server inside docker network.

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java (hunk 6 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: Do you mean put all code from partitioned methods under `start()` ?

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java (hunk 1 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: done

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java (hunk 3 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: I put 10s

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java (hunk 4 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/DockerComposeIT.java]: done

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java (hunk 6 above)
Comment (jeqo) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: yeap, explaining/organazing with comments instead of multiple methods.

Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: yeah. I also though about creation abstract class, and than extract all `@BeforeClass` actions to that abstract class.

**Comment by jeqo** [comment]
+1

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        SCHEMA_REGISTRY_INSIDE_DOCKER_ENV =
+                "http://" + schemaRegistry.getNetworkAliases().get(0) + ":8081";
+    }
+
+    private static void setJsonServer() {
+        jsonServer =
+                new GenericContainer("zhenik/json-server")
+                        .withExposedPorts(80)
+                        // all containers put in same network
+                        .withNetwork(kafka.getNetwork())
```
Comment (bsideup) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: I would highly recommend declaring a network explicitly (via `Network.newNetwork()`) instead of using Kafka's network.
In fact, after https://github.com/testcontainers/testcontainers-java/pull/1458 we may stop setting the network by default on Kafka's container

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                        .withExposedPorts(8080)
+                        .withNetwork(kafka.getNetwork())
+                        .withEnv("APPLICATION_PORT", "8080")
+                        .withEnv("KAFKA_BOOTSTRAP_SERVERS", KAFKA_BROKER_INSIDE_DOCKER_ENV)
+                        .withEnv("SCHEMA_REGISTRY_URL", SCHEMA_REGISTRY_INSIDE_DOCKER_ENV)
+                        .withEnv("SINK_TOPIC", TOPIC)
+                        .waitingFor(Wait.forHttp("/messages").forStatusCode(200));
+        httpProducer.start();
+
+        HTTP_PRODUCER_EXPOSED = LOCAL_HOST + ":" + httpProducer.getMappedPort(8080);
```
Comment (bsideup) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: please always use `getContainerIpAddress` instead of `localhost`, see https://www.testcontainers.org/features/networking/#getting-the-container-ip-address

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java (hunk 7 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: done

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java (hunk 8 above)
Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java]: done

File: e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersTestBaseIT.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    static String JSON_SERVER_INSIDE_DOCKER_ENV;
+    static String JSON_SERVER_EXPOSED;
+
+    static GenericContainer httpProducer;
+    static String HTTP_PRODUCER_EXPOSED;
+
+    static GenericContainer httpMaterializer;
+
+    @BeforeClass
+    public static void start() {
```
Comment (bsideup) [inline e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersTestBaseIT.java]: Since you extracted the base class, be aware that `@BeforeClass` will be triggered for **each** class that extends from this one. In other words, if you have 3 classes `extends ContainersTestBaseIT`, then you will create 3(!!!) networks, Kafka containers, and everything else. And, since you're not using `@AfterClass`, they will stay until the JVM exits.

Consider checking that `start` was already triggered, or use `static {}` block (aka "Singleton containers pattern": https://www.testcontainers.org/test_framework_integration/manual_lifecycle_control/#singleton-containers )

Comment (zhenik) [reply e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersTestBaseIT.java]: That is very useful, thx.
